### PR TITLE
std.complex: Add complex exp(), log(), log10(), and pow()

### PR DESCRIPTION
**Pull request by ibuclaw**
Pushing out initial implementation. Unittest incoming...

### CODE REVIEW

File: std/complex.d [hunk 1]
```diff
@@ -1047,3 +1047,104 @@ Complex!T sqrt(T)(Complex!T z)  @safe pure nothrow @nogc
...
+/**
+ * Calculates x$(SUPERSCRIPT n).
+ * The branch cut is on the negative axis.
+ * Params:
+ *      x = base
+ *      n = exponent
+ * Returns:
+ *      `x` raised to the power of `n`
+ */
+Complex!T pow(T, Int)(Complex!T x, Int n) @safe pure nothrow @nogc
```
Comment (n8sh) [inline std/complex.d:1103]:
```suggestion
Complex!T pow(T, Int)(Complex!T x, const Int n) @safe pure nothrow @nogc
```
For IFTI.

**Comment by ibuclaw** [comment]
Looks like tests are failing because std.complex doesn't have any template constraints for all it's functions.  Apparently indirect constraints aren't enough (`Complex!T foo(T)() { ... }` doesn't inherit constraints from it's return type `struct Complex(T) if (isFloatingPoint!T)`)

**Comment by ibuclaw** [comment]
Actually, I can't reproduce with what I thought would be a reduced test, so must be a bug in partial then.

File: std/complex.d [hunk 2]
```diff
@@ -1047,3 +1047,294 @@ Complex!T sqrt(T)(Complex!T z)  @safe pure nothrow @nogc
...
+
+    return fromPolar!(T, T)(std.math.exp(x.re), x.im);
+}
+
+///
+@safe pure nothrow @nogc unittest
+{
+    import std.math : PI;
+
+    assert(exp(complex(0.0, 1.0) * PI) == complex(1.0));
```
Comment (thewilsonator) [inline std/complex.d:1111]: Ummm...

File: std/complex.d [hunk 3]
```diff
@@ -1047,3 +1047,294 @@ Complex!T sqrt(T)(Complex!T z)  @safe pure nothrow @nogc
...
+            return Complex!T(T.nan, T.nan);
+        if (x.im == 0.0)
+            return x;
+    }
+    if (x.re == 0.0)
+    {
+        if (std.math.isNaN(x.im) || std.math.isInfinity(x.im))
+            return Complex!T(T.nan, T.nan);
+        else
+            return Complex!T(1.0, 0.0);
```
Comment (thewilsonator) [inline std/complex.d:1100]: there is still a phase you need to worry about here.

Comment (ibuclaw) [reply std/complex.d:1100]: Forgot the `if (x.im == 0.0)` bit. :-)

Comment (thewilsonator) [reply std/complex.d:1100]: Right, and fixing this should now cause https://github.com/dlang/phobos/pull/7456#discussion_r418970871 to fail.

File: std/complex.d (hunk 2 above)
Comment (ibuclaw) [reply std/complex.d:1111]: Yep, complex(-1) ... and needs approxEqual too. :-\

Comment (thewilsonator) [reply std/complex.d]: That is a pity that we need approxEqual for such a fundamental equality.

**Comment by ibuclaw** [comment]
Still WIP are:
- log() special cases.
- Put both exp() and log() special cases in Ddoc table.
- Changelog entry?

Best for another PR...
- tan is missing, as are also:
- asin, acos, atan
- sinh, cosh, tanh
- norm (squared magnitude)
- proj (projection onto the Riemann sphere)

File: std/complex.d [hunk 4]
```diff
@@ -1047,3 +1056,536 @@ Complex!T sqrt(T)(Complex!T z)  @safe pure nothrow @nogc
...
+@safe pure nothrow @nogc unittest
+{
+    import std.math : LN10, PI, sqrt;
+
+    auto a = complex(2.0, 1.0);
+    assert(log10(a) == log(a) / log(complex(10.0)));
+
+    auto b = complex(0.0, 1.0);
+    auto c = complex(sqrt(2.0) / 2, sqrt(2.0) / 2);
+    assert(log10(b) * 2.0 == log10(c) * 4.0);
```
Comment (ibuclaw) [inline std/complex.d:1381]: Still a test that fails on 32bits for "reasons"

**Comment by ibuclaw** [comment]
@thewilsonator - green after changing a few more tests from `==` to `approxEqual`.

These are a little tedious to write out, would it make sense to add an `approxEqual` helper for `Complex!T`?

**Comment by PetarKirov** [comment]
> would it make sense to add an approxEqual helper for Complex!T?

Or better yet, to `isClose` (https://github.com/dlang/phobos/pull/7384).

**Comment by ibuclaw** [comment]
Excellent, I better update all my tests then. :-)

**Comment by ibuclaw** [comment]
> Excellent, I better update all my tests then. :-)

That can wait for a follow-up PR anyhow, as most tests are isClose, but some only pass with approxEqual.

**Comment by ibuclaw** [comment]
@thewilsonator - anything left here?

**Pull request merged**